### PR TITLE
Use ReposourceCloneURLToRepoName in editor endpoint

### PR DESCRIPTION
**Pull request by felixfbecker**
The `/-/editor` endpoint currently just "guesses" the repo name, which is presumably the cause for https://github.com/sourcegraph/sourcegraph/issues/14423, because it for example doesn't take into account `repositoryPathPatterns`. Various editor-side config settings were proposed to work around this, but the best fix would be for the backend to just properly resolve clone URLs using its knowledge of external services.
We actually have good logic for resolving git clone URLs to repo names instead of guessing the repo name in the GraphQL API, but it was not used in this endpoint.
This PR moves that logic into a shared package and makes the editor endpoint use it too.

**Help needed**:
- We had no clue in which folder to put this shared function and just put it somewhere. There is probably a better place?
- The `TestEditorRedirect` test fails with a null derefence panic now. I would guess this is because one of those global mocks are missing, but since everything compiles and I don't know what dependencies are required, I am lost how to fix this.

@sourcegraph/cloud can a helpful backend person help us out?

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,8 +24,8 @@ All notable changes to Sourcegraph are documented in this file.
 ### Fixed
 
 - Imported changesets acquired an extra button to download the "generated diff", which did nothing, since imported changesets don't have a generated diff. This button has been removed. [#16778](https://github.com/sourcegraph/sourcegraph/issues/16778)
-
 - Quoted global filter values (case, patterntype) are now properly extracted and set in URL parameters. [#16186](https://github.com/sourcegraph/sourcegraph/issues/16186)
+- The endpoint for "Open in Sourcegraph" functionality in editor extensions now uses external service information to resolve the repository, which makes it more correct and respect the `repositoryPathPattern` setting. [#16846](https://github.com/sourcegraph/sourcegraph/pull/16846)
```
Comment (unknwon) [inline CHANGELOG.md:28]: nit:

```suggestion
- The endpoint for "Open in Sourcegraph" functionality in editor extensions now uses code host connection information to resolve the repository, which makes it more correct and respect the `repositoryPathPattern` setting. [#16846](https://github.com/sourcegraph/sourcegraph/pull/16846)
```

**Review by unknwon — APPROVED** [review]

**Comment by unknwon** [comment]
Thanks @felixfbecker! I've pushed some commits to add the missing mock, but I also see something unrelated failed, not sure if that's a cascading failure, see if CI passes this time.

**Comment by felixfbecker** [comment]
Thanks a lot @unknwon! So was the place where we put this function correct?

**Comment by unknwon** [comment]
>So was the place where we put this function correct?

I can't say if it's "correct", but at least seems reasonable to me.

**Pull request merged**